### ISSUE TITLE
Re-architect to avoid functorizing typechecker, evaluator

### ISSUE THREAD

**Issue by jonsterling** [RedPRL/redtt#162]
Right now, many parts of the core language are functorized over the global environment; it was done this way because I added the global environment only after I had already implemented this stuff, and I didn't want to add a parameter to every single function.

But it's kind of annoying and makes it difficult to program with, so I would like to consider a more basic architecture, where we have to either add an argument, or even maybe use a monad.

I'm not completely decided whether we should do this, but I wanted to open it to discussion.

**Comment by favonia** [comment]
Is this related to the urgently needed rewrite of `Tm.ml`?

**Comment by jonsterling** [comment]
@favonia I think it's a separate issue, as `Tm.ml` isn't functorized.

**Comment by jonsterling** [comment]
But generally, they both fall under the category of "eliminating cruft".

**Comment by jonsterling** [comment]
@favonia Another thing I'd like to do is present the typechecker as a kind of machine, which emits a collection of constraints and blocked subgoals if you run it to quiescence. A blocked subgoal is something that might typecheck, if a unification constraint could be solved.

This might be a nice idea, because if still lets us keep unification out of the core, but it lets us see *why* typechecking failed, and what we need to do in order to make it succeed. I envision running the typechecker, feeding the resulting constraints into the unifier, and then *resuming* the blocked subgoals after unification has succeeded.

**Comment by jonsterling** [comment]
(This is, by the way, essential for unleashing the immortal algorithmic type system with eta-respecting subtyping. Otherwise, we'd need to implement the typechecker once in the core, and again in the unifier.)

**Comment by jonsterling** [comment]
Related to this issue is the fact that the current style, where the "global context" remains constant throughout typechecking, is incompatible with the modal calculus I want to develop. We really do need to pass around that context and environment everywhere, so that we have the opportunity to modify it in certain important places.

**Comment by jonsterling** [comment]
In #177 I've resolved the issue where the global context couldn't be locally updated by the typechecker. We are also now closer to being able to do this proposed re-architecture.
